app/components/one_time_code_input_component.rb

```diff
@@ -53,11 +60,23 @@ def hint
     end
   end
 
+  def input_maxlength
+    optional_prefix.size + code_length
+  end
+
   def input_pattern
+    "#{input_pattern_prefix}#{input_pattern_character_set}{#{code_length}}"
+  end
+
+  def input_pattern_prefix
+    "#{regexp_escape_for_js(optional_prefix)}?" if optional_prefix.present?
+  end
+
+  def input_pattern_character_set
     if numeric?
-      '#?[0-9]*'
+      '[0-9]'
     else
-      '#?[a-zA-Z0-9]*'
+      '[a-z0-9]'
```

Comment:
Whoops, made a bold (incorrect) assumption that the pattern would be case-insensitive by default. Luckily specs caught me on it!

```suggestion
      '[a-zA-Z0-9]'
```